Go tutorial and reference (#723)
canonical/rockcraft: docs/tutorial/go.rst

install ``go``.

.. literalinclude:: code/go/task.yaml
    :language: bash
    :start-after: [docs:install-go]
    :end-before: [docs:install-go-end]
    :dedent: 2

Create a ``main.go`` file, copy the following text into it and then
save it:

.. literalinclude:: code/go/main.go
    :language: go

Initialise the new go module:

.. literalinclude:: code/go/task.yaml
    :language: bash
    :start-after: [docs:mod-init-go]
    :end-before: [docs:mod-init-go-end]
    :dedent: 2

Build the Go application so it can be run:

.. literalinclude:: code/go/task.yaml
    :language: bash
    :start-after: [docs:go-build]
    :end-before: [docs:go-build-end]
    :dedent: 2

A new binary named ``go-hello-world`` should be in the current directory.
Run the Go application using ``./go-hello-world`` to verify that
it works.

Test the Go application by using ``curl`` to send a request to the root
endpoint. You may need a new terminal for this, if you are using Multipass use
``multipass shell rock-dev`` to get another terminal:

.. literalinclude:: code/go/task.yaml
    :language: bash
    :start-after: [docs:curl-go]
    :end-before: [docs:curl-go-end]
    :dedent: 2

> The Go application should respond with ``Hello World``.

Reviewer: I think this needs to be updated

    you are running on an ``amd64`` platform. Check the architecture of your
    system using ``dpkg --print-architecture``.

    The ``name``, ``version`` and ``platform`` all influence the name of the
    generated ``.rock`` file.

Pack the rock:

.. literalinclude:: code/go/task.yaml
    :language: bash
    :start-after: [docs:pack]
    :end-before: [docs:pack-end]
    :dedent: 2

.. note::

    Depending on your network, this step can take a couple of minutes to finish.

    ``ROCKCRAFT_ENABLE_EXPERIMENTAL_EXTENSIONS`` is required whilst the Go
    extension is experimental.

Once Rockcraft has finished packing the Go rock, you'll find a new file in
your working directory (an `OCI <OCI_image_spec_>`_ archive) with the ``.rock``
extension:

.. literalinclude:: code/go/task.yaml
    :language: bash
    :start-after: [docs:ls-rock]
    :end-before: [docs:ls-rock-end]
    :dedent: 2

.. note::
    If you changed the ``name`` or ``version`` in ``rockcraft.yaml`` or are not
    on an ``amd64`` platform, the name of the ``.rock`` file will be different
    for you.

    By default, the ``go-framework`` will use the ``bare`` base. You can improve

>     the developer experience changing the base to ``ubuntu@24.04``, but the

Reviewer: Could we caveat what we mean by this? E.g., for debugging it might be better to use but for production we recommend bare?

.. _build-a-rock-for-a-go-application:

Build a rock for a Go application
------------------------------------

> In this tutorial, we'll create a simple Go application and learn how to

Author: I have updated it so the subject is always we (or the sentences are in imperative mode).

It appropriate, we may need to update the rest of the tutorial @jdkandersson .

> In this tutorial, we'll create a simple Go application and learn how to

Reviewer: Yes we should do that please

> In this tutorial, we'll create a simple Go application and learn how to

Reviewer 2: I'll be opening a PR soon to migrate some of the Charm SDK docs into Rockcraft; I'll also look over the tutorials and update them to change the subject to we or use imperative mode.

containerise it in a rock, using Rockcraft's ``go-framework``
:ref:`extension <go-framework-reference>`.

Setup
=====

> .. include:: /reuse/tutorial/setup.rst

Reviewer: I think it is a good point, we should remove the reference to which version of Ubuntu we start with from the shared setup

> .. include:: /reuse/tutorial/setup.rst

Author: Yeah, I have removed the version of Ubuntu, now it is something like:
```
we recommend starting from a clean Ubuntu installation. If we don't have one available, we can create one using Multipass
```
so also it it a "bit" clearer than Multipass is not required, and the requirement is Ubuntu. By the way, one of the reasons for Multipass is to be able to run the tutorial in macOS/Windows operating systems (rockcraft requires a Linux distribution with snapd installed).

> .. include:: /reuse/tutorial/setup.rst

Author: We recommend a clean installation (clearly, it is not a requirement). Frankly, until there are issues with docker, I think it is a good recommendation...

.. code-block:: bash

   mkdir go-hello-world
   cd go-hello-world

Create the Go application
==============================

Start by creating the "Hello, world" Go application that will be used for
this tutorial.

In order to test the Go application locally (before packing it into a rock),
install ``go``.

.. literalinclude:: code/go/task.yaml
    :language: bash
    :start-after: [docs:install-go]
    :end-before: [docs:install-go-end]
    :dedent: 2

Create a ``main.go`` file, copy the following text into it and then
save it:

.. literalinclude:: code/go/main.go
    :language: go

Initialise the new go module:

.. literalinclude:: code/go/task.yaml
    :language: bash
    :start-after: [docs:mod-init-go]
    :end-before: [docs:mod-init-go-end]
    :dedent: 2

Build the Go application so it can be run:

.. literalinclude:: code/go/task.yaml
    :language: bash
    :start-after: [docs:go-build]
    :end-before: [docs:go-build-end]
    :dedent: 2

A new binary named ``go-hello-world`` should be in the current directory.
Run the Go application using ``./go-hello-world`` to verify that
it works.

Test the Go application by using ``curl`` to send a request to the root

> endpoint. You may need a new terminal for this, if you are using Multipass use

Reviewer: I think the reason I originally chose `may` rather than `will` is because there are other ways a user could test this. I do agree we can probably just write `will` here, and can also fix up the rest of the tutorials

On Multipass, it is not guaranteed they are, we suggest using it

> endpoint. You may need a new terminal for this, if you are using Multipass use

Author: Resolving this conversation. Please, reopen it it you think it is still relevant.

``multipass shell rock-dev`` to get another terminal:

.. literalinclude:: code/go/task.yaml
    :language: bash
    :start-after: [docs:curl-go]
    :end-before: [docs:curl-go-end]
    :dedent: 2

The Go application should respond with ``Hello, world!``.

The Go application looks good, so stop it for now using :kbd:`Ctrl` + :kbd:`C`.

Pack the Go application into a rock
===================================

First, we'll need a ``rockcraft.yaml`` file. Rockcraft will automate its
creation and tailoring for a Go application by using the

> ``go`` profile:

Reviewer: Yeh this should be `go-framework`

.. literalinclude:: code/go/task.yaml
    :language: bash
    :start-after: [docs:create-rockcraft-yaml]
    :end-before: [docs:create-rockcraft-yaml-end]
    :dedent: 2

The ``rockcraft.yaml`` file will automatically be created in your working

> directory. Open it in a text editor and check that the ``name`` is

Reviewer: I'm hesitant in showing the file because then we have to keep it up to date with the template. We are explicit about what we say the user should check so I think it should be ok

``go-hello-world``. Ensure that ``platforms`` includes the architecture of
your host. For example, if your host uses the ARM
architecture, include ``arm64`` in ``platforms``.

.. note::
    For this tutorial, we'll use the ``name`` ``go-hello-world`` and assume
    you are running on an ``amd64`` platform. Check the architecture of your
    system using ``dpkg --print-architecture``.

    The ``name``, ``version`` and ``platform`` all influence the name of the
    generated ``.rock`` file.

Pack the rock:

.. literalinclude:: code/go/task.yaml
    :language: bash
    :start-after: [docs:pack]
    :end-before: [docs:pack-end]
    :dedent: 2

.. note::

    Depending on your network, this step can take a couple of minutes to finish.

    ``ROCKCRAFT_ENABLE_EXPERIMENTAL_EXTENSIONS`` is required whilst the Go

>     extension is experimental.

Reviewer: Agreed, should be `go-framework`

Once Rockcraft has finished packing the Go rock, you'll find a new file in
your working directory (an `OCI <OCI_image_spec_>`_ archive) with the ``.rock``
extension:

.. literalinclude:: code/go/task.yaml
    :language: bash
    :start-after: [docs:ls-rock]
    :end-before: [docs:ls-rock-end]
    :dedent: 2

.. note::
    If you changed the ``name`` or ``version`` in ``rockcraft.yaml`` or are not
    on an ``amd64`` platform, the name of the ``.rock`` file will be different
    for you.

    By default, the ``go-framework`` will use the ``bare`` base.
    The ``bare`` base is recommended for production applications because
    the image size will be smaller, and the risk of security issues lower.
    However, you may want to change the base to ``ubuntu@24.04`` during the
    development process to benefit from the extra utilities provided.

> Run the Go rock with Docker

Reviewer: Sounds reasonable, we can probably do that in the shared setup instructions

> Run the Go rock with Docker

Author: I have updated to use docker as a regular user. I have not updated the rest of tutorials (basically all of them). I would rather do it in another PR.

    :end-before: [docs:docker-images-end]
    :dedent: 2

The output should list the Go container image, along with its tag, ID and
size:

..  code-block:: text
    :class: log-snippets

    REPOSITORY       TAG       IMAGE ID       CREATED         SIZE
    go-hello-world   0.1       f3abf7ebc169   5 minutes ago   15.7MB

.. note::
    The size of the image reported by Docker is the uncompressed size which is
    larger than the size of the compressed ``.rock`` file.

Now we're finally ready to run the rock and test your containerised Go
application:

.. literalinclude:: code/go/task.yaml
    :language: text
    :start-after: [docs:docker-run]
    :end-before: [docs:docker-run-end]
    :dedent: 2

Use the same ``curl`` command as before to send a request to the Go
application's root endpoint which is running inside the container:

.. literalinclude:: code/go/task.yaml
    :language: text
    :start-after: [docs:curl-go-rock]
    :end-before: [docs:curl-go-rock-end]
    :dedent: 2

The Go application should again respond with ``Hello, world!``.

View the application logs
~~~~~~~~~~~~~~~~~~~~~~~~~

When deploying the Go rock, you can always get the application logs via

> ``pebble``:

Reviewer 3: This is the very first occurrence of "pebble" in the whole rockcraft doc website: It doesn't show up on the home page, it doesn't show up in the first two tutorials, either.

Following the doc's structure, I would read the homepage, follow the first two tutorials, then follow the Go tutorial because other languages don't interest me, and bam, I met the term "pebble", and it's confusing: What is pebble?

Although there is a link, but again, this adds cognitive load for a tutorial that is supposed to be simple, as simple as building and running an OCI image.

Even if the user is already familiar with pebble or switched to the pebble doc and then back, they would still have a question mark: I already built an executable binary, why can't I just add it in the OCI image and run it, why would I need a service manager for a single executable?

Reviewer: I think we can link to information about pebble, I don't think we should get into the details of explaining pebble here. We should update the rest of the tutorials as well

Author: I updated the tutorial so the first reference to pebble has the link. I do not think pebble will be a hurdle for the user here, as the user can just gloss over it (and run the command and see the logs). It they want to have more information about it, there is no other options except reading the docs (and explaining it here will make the tutorial complex).

.. literalinclude:: code/go/task.yaml
    :language: text
    :start-after: [docs:get-logs]
    :end-before: [docs:get-logs-end]
    :dedent: 2

As a result, :ref:`pebble_explanation_page` will give you the logs for the
``go`` service running inside the container.
You should expect to see something similar to this:

..  code-block:: text
    :class: log-snippets

    2024-10-04T08:51:35.826Z [go] 2024/10/04 08:51:35 starting hello world application
    2024-10-04T08:51:39.974Z [go] 2024/10/04 08:51:39 new hello world request

You can also choose to follow the logs by using the ``-f`` option with the
``pebble logs`` command above. To stop following the logs, press :kbd:`Ctrl` + :kbd:`C`.

Cleanup
~~~~~~~

> Now we have a fully functional rock for a Go application! This concludes

Reviewer 3: I find the cleanup section in the middle of a tutorial a little weird. I read "cleanup" then immediately "update go application" -- if I'm still working on the app, why do I need to clean it up now? The next step builds the image with a different tag anyway. I think it's better to keep everything that is cleanup together at the end, simply delete the multipass VM which is created in the setup section. And we can make the cleanup section as a reusable content and use it in other tutorials as well.

Reviewer: The reason is that when we run the next image it will use the same port. I do agree we should just have this without necessarily a cleanup title and just stop the container and remove the image. We should update the rest of the tutorials as well

Author: I have updated the title to be "stop the application" instead of "cleanup". We need a title as the previous one is "View the application logs". The alternative is to put it into the "update go application".

    Note that the resulting ``.rock`` file will now be named differently, as
    its new version will be part of the filename.

Finally, use ``curl`` to send a request to the ``/time`` endpoint:

.. literalinclude:: code/go/task.yaml
    :language: text
    :start-after: [docs:curl-time]
    :end-before: [docs:curl-time-end]
    :dedent: 2

The updated application should respond with the current date and time (e.g.
``2024-06-21 09:47:56``).

.. note::

    If you are not getting the current date and time from the ``/time``
    endpoint, check the :ref:`troubleshooting-go` steps below.

Cleanup
~~~~~~~

We can now stop the container and remove the corresponding image:

.. literalinclude:: code/go/task.yaml
    :language: bash
    :start-after: [docs:stop-docker-updated]
    :end-before: [docs:stop-docker-updated-end]
    :dedent: 2

Reset your environment
======================

We've reached the end of this tutorial.

If you'd like to reset your working environment, you can simply run the
following:

.. literalinclude:: code/go/task.yaml
    :language: bash
    :start-after: [docs:cleanup]
    :end-before: [docs:cleanup-end]
    :dedent: 2

.. collapse:: If using Multipass...

>     If you created an instance using Multipass, you can also clean it up.

Reviewer 3: "If you created an instance using Multipass" -- I did create an instance with multipass, it's the only option provided in the setup section. Is there another option? If so, maybe we can enrich the setup part. If not, we can reword this, simply delete the VM is fine?

Reviewer: I think it depends, some users don't want to use multipass

    Start by exiting it:

    .. code-block:: bash

        exit

    And then you can proceed with its deletion:

    .. code-block:: bash

        multipass delete rock-dev
        multipass purge

----

.. _troubleshooting-go:

Troubleshooting
===============

**Application updates not taking effect?**

Upon changing your Go application and re-packing the rock, if you believe
your changes are not taking effect, try running ``rockcraft clean`` and pack
the rock again with ``rockcraft pack``.

> 

Reviewer 3: Maybe in the end, we can add something like "what's next", "where to go from here", to add more links for the users, like links to explanations and how-tos.

Reviewer: Sounds reasonable, we should do that for the rest of the tutorials as well

Author: I have added a section "next steps" (following the juju docs nomenclature). Let's discuss what to put in here for this tutorial and do it for the rest of the tutorials in another PR.

    :language: bash
    :start-after: [docs:install-go]
    :end-before: [docs:install-go-end]
    :dedent: 2

Initialise the new Go module:

.. literalinclude:: code/go/task.yaml
    :language: bash
    :start-after: [docs:mod-init-go]
    :end-before: [docs:mod-init-go-end]
    :dedent: 2

Create a ``main.go`` file, copy the following text into it and then
save it:

.. literalinclude:: code/go/main.go
    :language: go

Build the Go application so it can be run:

.. literalinclude:: code/go/task.yaml
    :language: bash
    :start-after: [docs:go-build]
    :end-before: [docs:go-build-end]
    :dedent: 2

A new binary named ``go-hello-world`` should be in the current directory.
Run the Go application using ``./go-hello-world`` to verify that
it works.

Test the Go application by using ``curl`` to send a request to the root
endpoint. We may need a new terminal for this, if we are using Multipass use
``multipass shell rock-dev`` to get another terminal:

.. literalinclude:: code/go/task.yaml
    :language: bash
    :start-after: [docs:curl-go]
    :end-before: [docs:curl-go-end]
    :dedent: 2

The Go application should respond with ``Hello, world!``.

> The Go application looks good, so stop it for now using :kbd:`Ctrl` + :kbd:`C`.

Reviewer 4: I naively thought I needed to keep the binary present in the project dir. :laughing: 
Maybe add a comment that we can delete the binary or we don't need it?

Author: Resolving this conversation. Please, reopen it it you think it is still relevant.

Pack the Go application into a rock
===================================

First, we'll need a ``rockcraft.yaml`` file. Rockcraft will automate its
creation and tailor it for a Go application by using the
``go-framework`` profile:

.. literalinclude:: code/go/task.yaml
    :language: bash
    :start-after: [docs:create-rockcraft-yaml]
    :end-before: [docs:create-rockcraft-yaml-end]
    :dedent: 2

The ``rockcraft.yaml`` file will automatically be created in the project working
directory. Open it in a text editor and check that the ``name`` is

> ``go-hello-world``. Ensure that ``platforms`` includes the architecture of

Author: In the charmcraft tutorials, we explicitly set the name of the charm with the `--name` option (as in that case the directory is named `charm`), however, in the rockcraft tutorials, we expect the user has followed the previous steps and the name is correct. That part `Open it in a text editor and check that the`.... is in all tutorials, and I would leave it that way, as I believe the intention is that the user just opens and checks the file (but the name should be as stated).

We should expect to see something similar to this:

..  code-block:: text
    :class: log-snippets

    2024-10-04T08:51:35.826Z [go] 2024/10/04 08:51:35 starting hello world application
    2024-10-04T08:51:39.974Z [go] 2024/10/04 08:51:39 new hello world request

We can also choose to follow the logs by using the ``-f`` option with the
``pebble logs`` command above. To stop following the logs, press :kbd:`Ctrl` + :kbd:`C`.

Stop the application
~~~~~~~~~~~~~~~~~~~~

Now we have a fully functional rock for a Go application! This concludes
the first part of this tutorial, so we'll stop the container and remove the
respective image for now:

.. literalinclude:: code/go/task.yaml
    :language: bash
    :start-after: [docs:stop-docker]
    :end-before: [docs:stop-docker-end]
    :dedent: 2

Update Go application
========================================

As a final step, let's update our application. For example,
we want to add a new ``/time`` endpoint which returns the current time.

Start by opening the ``main.go`` file in a text editor and update the code to
look like the following:

.. literalinclude:: code/go/main.go.time
    :language: go

> Since we are creating a new version of the application, **open the**

Reviewer 5: ```suggestion
Since we are creating a new version of the application, open the
```

   mkdir go-hello-world
   cd go-hello-world

Create the Go application
==============================

Start by creating the "Hello, world" Go application that will be used for
this tutorial.

In order to test the Go application locally (before packing it into a rock),
install ``go``.

.. literalinclude:: code/go/task.yaml
    :language: bash
    :start-after: [docs:install-go]
    :end-before: [docs:install-go-end]
    :dedent: 2

Initialise the new Go module:

.. literalinclude:: code/go/task.yaml
    :language: bash
    :start-after: [docs:mod-init-go]
    :end-before: [docs:mod-init-go-end]
    :dedent: 2

Create a ``main.go`` file, copy the following text into it and then
save it:

.. literalinclude:: code/go/main.go
    :language: go

Build the Go application so it can be run:

.. literalinclude:: code/go/task.yaml
    :language: bash
    :start-after: [docs:go-build]
    :end-before: [docs:go-build-end]
    :dedent: 2

A new binary named ``go-hello-world`` should be in the current
directory.  This binary is only needed for local testing, as
Rockcraft will compile the Go application. Run the Go application using
``./go-hello-world`` to verify that it works.

> Test the Go application by using ``curl`` to send a request to the root

Reviewer 5: I feel like there's a jump here because 'hello world' code in Go does not typically start up an http server. Maybe some context like "The code in ``main.go`` starts up an HTTP server listening on port ``8000``, so we test it by using ``curl``..."

.. literalinclude:: code/go/task.yaml
    :language: bash
    :start-after: [docs:curl-go]
    :end-before: [docs:curl-go-end]
    :dedent: 2

The Go application should respond with ``Hello, world!``.

The Go application looks good, so stop it for now using :kbd:`Ctrl` + :kbd:`C`.

Pack the Go application into a rock
===================================

First, we'll need a ``rockcraft.yaml`` file. Rockcraft will automate its
creation and tailor it for a Go application by using the
``go-framework`` profile:

.. literalinclude:: code/go/task.yaml
    :language: bash
    :start-after: [docs:create-rockcraft-yaml]
    :end-before: [docs:create-rockcraft-yaml-end]
    :dedent: 2

The ``rockcraft.yaml`` file will automatically be created in the project working
directory. Open it in a text editor and check that the ``name`` is
``go-hello-world``. Ensure that ``platforms`` includes the architecture of
the host. For example, if the host uses the ARM
architecture, include ``arm64`` in ``platforms``.

.. note::
    For this tutorial, we'll use the ``name`` ``go-hello-world`` and assume
    we are running on an ``amd64`` platform. Check the architecture of the

>     system using ``dpkg --print-architecture``.

Author: Up to this point no rock was generated, so the user at this point has not seen a file name different to the tutorial. It is in the output of the following command `rockcraft pack` when the first discrepancy could arise. Not sure we should put it before, what do you think?

>     system using ``dpkg --print-architecture``.

Reviewer 6: I think the bigger issue is that the `setup.rst` template we're reusing doesn't leave us room to narrow the focus to a particular arch. Not your problem to solve in this PR, so I'm resolving this. :)

    The ``name``, ``version`` and ``platform`` all influence the name of the
    generated ``.rock`` file.

Pack the rock:

.. literalinclude:: code/go/task.yaml
    :language: bash
    :start-after: [docs:pack]
    :end-before: [docs:pack-end]
    :dedent: 2

.. note::

    Depending on the network, this step can take a couple of minutes to finish.

    ``ROCKCRAFT_ENABLE_EXPERIMENTAL_EXTENSIONS`` is required whilst the ``go-framework``

>     extension is experimental.

Reviewer 6: If this is still accurate, then we urgently need to make this a separate step. Dirtying the `rockcraft pack` step with a shell var declaration diminishes its importance.

Author: I have updated the tutorial so the exported envvar `ROCKCRAFT_ENABLE_EXPERIMENTAL_EXTENSIONS` is added in the first part of the tutorial, so it is not added to the `rockcraft pack` commands.

Reviewer 6: Looks good, thanks!

> .. note::
> 
>     Depending on the network, this step can take a couple of minutes to finish.
> 
>     ``ROCKCRAFT_ENABLE_EXPERIMENTAL_EXTENSIONS`` is required whilst the ``go-framework``
>     extension is experimental.

Reviewer 6: ```suggestion
Depending on the network, this step can take a couple of minutes to finish.

``ROCKCRAFT_ENABLE_EXPERIMENTAL_EXTENSIONS`` is required while the ``go-framework``
extension is experimental.
```

Once Rockcraft has finished packing the Go rock, we'll find a new file in

> the working directory (an `OCI <OCI_image_spec_>`_ archive) with the ``.rock``

Reviewer 6: > archive

Wouldn't "image" be more appropriate? Or are we avoiding that word?

extension:

.. literalinclude:: code/go/task.yaml
    :language: bash
    :start-after: [docs:ls-rock]
    :end-before: [docs:ls-rock-end]
    :dedent: 2

.. note::
    If we changed the ``name`` or ``version`` in ``rockcraft.yaml`` or are not
    on an ``amd64`` platform, the name of the ``.rock`` file will be different.

    By default, the ``go-framework`` will use the ``bare`` base.
    The ``bare`` base is recommended for production applications because
    the image size will be smaller, and the risk of security issues lower.
    However, we may want to change the base to ``ubuntu@24.04`` during the

>     development process to benefit from the extra utilities provided.

Reviewer 6: This is all supplemental information more suited toward a how-to or reference, especially when we consider that later we specifically mention the resulting rock name as `go-hello-world_0.1_amd64.rock`.

If you're amenable, I'd like to remove it.

Author: sure, removed the full note.

Run the Go rock with Docker
===========================

We already have the rock as an `OCI <OCI_image_spec_>`_ archive. Load the
image into Docker:

.. literalinclude:: code/go/task.yaml
    :language: bash
    :start-after: [docs:skopeo-copy]
    :end-before: [docs:skopeo-copy-end]
    :dedent: 2

Check that the image was successfully loaded into Docker:

.. literalinclude:: code/go/task.yaml
    :language: bash
    :start-after: [docs:docker-images]
    :end-before: [docs:docker-images-end]
    :dedent: 2

The output should list the Go container image, along with its tag, ID and
size:

..  code-block:: text
    :class: log-snippets

    REPOSITORY       TAG       IMAGE ID       CREATED         SIZE
    go-hello-world   0.1       f3abf7ebc169   5 minutes ago   15.7MB

.. note::
    The size of the image reported by Docker is the uncompressed size which is

>     larger than the size of the compressed ``.rock`` file.

Reviewer 6: Can we cut this? I don't think it adds much.

Now we're finally ready to run the rock and test the containerised Go
application:

.. literalinclude:: code/go/task.yaml
    :language: text
    :start-after: [docs:docker-run]
    :end-before: [docs:docker-run-end]
    :dedent: 2

Use the same ``curl`` command as before to send a request to the Go
application's root endpoint which is running inside the container:

.. literalinclude:: code/go/task.yaml
    :language: text
    :start-after: [docs:curl-go-rock]
    :end-before: [docs:curl-go-rock-end]
    :dedent: 2

> The Go application should again respond with ``Hello, world!``.

Reviewer 6: ```suggestion
The Go application again responds with ``Hello, world!``.
```

View the application logs
~~~~~~~~~~~~~~~~~~~~~~~~~

> When deploying the Go rock, we can always get the application logs via

Reviewer 6: ```suggestion
When deploying the Go rock, we can always get the application logs with
```

We must [avoid Latin](https://docs.ubuntu.com/styleguide/en/#latin-words-and-phrases).

.. literalinclude:: code/go/task.yaml
    :language: text
    :start-after: [docs:get-logs]
    :end-before: [docs:get-logs-end]
    :dedent: 2

As a result, :ref:`pebble_explanation_page` will give you the logs for the
``go`` service running inside the container.
You should expect to see something similar to this:

..  code-block:: text
    :class: log-snippets

    2024-10-01T06:32:50.180Z [fastapi] INFO:     Started server process [12]
    2024-10-01T06:32:50.181Z [fastapi] INFO:     Waiting for application startup.
    2024-10-01T06:32:50.181Z [fastapi] INFO:     Application startup complete.
    2024-10-01T06:32:50.182Z [fastapi] INFO:     Uvicorn running on http://0.0.0.0:8000 (Press CTRL+C to quit)

>     2024-10-01T06:32:58.214Z [fastapi] INFO:     172.17.0.1:55232 - "GET / HTTP/1.1" 200 OK

Author: :flushed:

You can also choose to follow the logs by using the ``-f`` option with the
``pebble logs`` command above. To stop following the logs, press :kbd:`Ctrl` + :kbd:`C`.

Cleanup
~~~~~~~

Now we have a fully functional rock for a Go application! This concludes
the first part of this tutorial, so we'll stop the container and remove the
respective image for now:

.. literalinclude:: code/go/task.yaml
    :language: bash
    :start-after: [docs:stop-docker]
    :end-before: [docs:stop-docker-end]
    :dedent: 2

> Add another binary to the Go application

Author: replied in the next comment.

========================================

As a final step, let's update our application to include another binary.

Create a new ``main.go`` file in the directory  ``cmd/anotherserver`` relative
to the base directory of the project, with the following text and then
save it:

.. literalinclude:: code/go/cmd/anotherserver/main.go
    :language: go

Add the following snippet to the file ``rockcraft.yaml``, so the new binary is

> set as the main application to run:

Author: I did this because I think it is one of the pain points of the extension. Many Go projects have the binaries under cmd/, sometimes being one the server and other ones could be for example command lines tools. 

I put this example to show how to change that, and at the same time how to "update" the application, put another version of the rock... Would you suggest another different example to update the app?

> set as the main application to run:

Reviewer: Perhaps that can be a separate how to. I would suggest just showing perhaps adding another endpoint like the time endpoint on other tutorials

> set as the main application to run:

Reviewer: I think the tutorial doesn’t flow, it doesn’t make sense at least to me that I would want to add another binary at this point, I probably would just want to add some features

> set as the main application to run:

Author: sure, let me try that way (adding another endpoint) and I will leave that part for a how-to (I think it is important)

> set as the main application to run:

Author: Done!

> ``rockcraft.yaml`` file and change the ``version`` (e.g. to ``0.2``).

Reviewer 6: ```suggestion
``rockcraft.yaml`` file and set ``version: "0.2"``.
```

.. literalinclude:: code/go/anotherserver_part
    :language: yaml

Since we are creating a new version of the application, **open the**
``rockcraft.yaml`` file and change the ``version`` (e.g. to ``0.2``).

.. note::

>     ``rockcraft pack`` will create a new image with the updated code even if you

Reviewer 2: Nitpick: Should this pack command have `ROCKCRAFT_ENABLE_EXPERIMENTAL_EXTENSIONS` included in it?

Author: Not sure about that, the command to run is just below this line (with the correct env var set). I think this is referring to the generic "rockcraft pack". What do you think?

    :start-after: [docs:docker-run-update]
    :end-before: [docs:docker-run-update-end]
    :dedent: 2

.. note::

    Note that the resulting ``.rock`` file will now be named differently, as
    its new version will be part of the filename.

Finally, use ``curl`` to send a request to the ``/time`` endpoint:

.. literalinclude:: code/go/task.yaml
    :language: text
    :start-after: [docs:curl-time]
    :end-before: [docs:curl-time-end]
    :dedent: 2

> The updated application should respond with the current date and time (e.g.
> ``2024-06-21 09:47:56``).

Reviewer 6: ```suggestion
The updated application will respond with the current date and time.
```

    :language: bash
    :start-after: [docs:stop-docker-updated]
    :end-before: [docs:stop-docker-updated-end]
    :dedent: 2

Reset the environment
======================

We've reached the end of this tutorial.

If we'd like to reset the working environment, we can simply run the
following:

.. literalinclude:: code/go/task.yaml
    :language: bash
    :start-after: [docs:cleanup]
    :end-before: [docs:cleanup-end]
    :dedent: 2

.. collapse:: If using Multipass...

    If we created an instance using Multipass, we can also clean it up.
    Start by exiting it:

    .. code-block:: bash

        exit

    And then we can proceed with its deletion:

    .. code-block:: bash

        multipass delete rock-dev
        multipass purge

----

Next steps
==========

* :ref:`Rockcraft tutorials<tutorial>`.
* :ref:`go-framework reference<go-framework-reference>`.
* :ref:`why_use_rockcraft`.
* :ref:`What is a Rock?<rocks_explanation>`.

----

.. _troubleshooting-go:

Troubleshooting
===============

**Application updates not taking effect?**

Upon changing the Go application and re-packing the rock, if
the changes are not taking effect, try running ``rockcraft clean`` and pack

> the rock again with ``rockcraft pack``.

Reviewer 6: Instead of linking here, I feel it would be better to simply instruct the user to run `rockcraft clean` prior to repacking the rock and remove this section. Even if `clean` isn't always required, it helps the reader build muscle memory of the commands they'll often need to reach for when iteratively building and testing.

.. _build-a-rock-for-a-go-application:

Build a rock for a Go application
------------------------------------

In this tutorial, we'll containerise a simple Go application into a rock using
Rockcraft's ``go-framework`` :ref:`extension <go-framework-reference>`.

It should take 25 minutes for you to complete.

You won’t need to come prepared with intricate knowledge of software
packaging, but familiarity with Linux paradigms, terminal operations,

> and ``go`` is required.
> 
> Once you complete this tutorial, you’ll have a working ``rock`` for a
> ``go`` application. You’ll gain familiarity with Rockcraft and the
> ``go-framework`` extension, and have the experience to create
> ``rocks`` for ``go`` applications.

Reviewer 6: ```suggestion
and Go is required.

Once you complete this tutorial, you’ll have a working rock for a
Go application. You’ll gain familiarity with Rockcraft and the
``go-framework`` extension, and have the experience to create
rocks for Go applications.
```

~~~~~~~~~~~~~~~~~~~~~~~~~

When deploying the Go rock, we can always get the application logs with
:ref:`pebble_explanation_page`:

.. literalinclude:: code/go/task.yaml
    :language: text
    :start-after: [docs:get-logs]
    :end-before: [docs:get-logs-end]
    :dedent: 2

As a result, ``pebble`` will give the logs for the
``go`` service running inside the container.
We should expect to see something similar to this:

..  code-block:: text
    :class: log-snippets

    2024-10-04T08:51:35.826Z [go] 2024/10/04 08:51:35 starting hello world application
    2024-10-04T08:51:39.974Z [go] 2024/10/04 08:51:39 new hello world request

We can also choose to follow the logs by using the ``-f`` option with the
``pebble logs`` command above. To stop following the logs, press :kbd:`Ctrl` + :kbd:`C`.

Stop the application
~~~~~~~~~~~~~~~~~~~~

Now we have a fully functional rock for a Go application! This concludes
the first part of this tutorial, so we'll stop the container and remove the
respective image for now:

.. literalinclude:: code/go/task.yaml
    :language: bash
    :start-after: [docs:stop-docker]
    :end-before: [docs:stop-docker-end]
    :dedent: 2

> Update Go application

Reviewer 6: ```suggestion
Update the Go application
```
